lyrics-fetcher-neteasecloud.el

```diff
@@ -0,0 +1,216 @@
+;; Maintainer: Korytov Pavel <thexcloud@gmail.com>
+;; Homepage: https://github.com/SqrtMinusOne/lyrics-fetcher.el
+
+;; This file is NOT part of GNU Emacs.
+
+;; This program is free software: you can redistribute it and/or modify
+;; it under the terms of the GNU General Public License as published by
+;; the Free Software Foundation, either version 3 of the License, or
+;; (at your option) any later version.
+
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; Fetch song lyrics from genius.com.
+
+;;; Code:
+(require 'request)
+(require 'cl-lib)
+(require 'json)
+(require 'seq)
+(require 'shr)
+(require 'f)
+(require 'dom)
+
+(defcustom lyrics-fetcher-neteasecloud-strip-parens-from-query t
+  "Strip parens from the query.
+
+I've noticed that these often break the search, e.g. when
+searching \"Song (feat. Artist)\""
+  :type 'boolean
+  :group 'lyrics-fetcher)
+
+(defun lyrics-fetcher-neteasecloud-do-search (track callback &optional sync edit)
+  "Perform a lyrics search on 'music.163.com'.
+
+The flow is as follows:
+1. Send a POST /search request with a text query
+2. Pick the first result (or prompt user if SYNC is non-nil)
+3. Fetch lyrics
+4. Call CALLBACK with the resulting lyrics string
+
+TRACK should be EMMS-compatible alist or string, take a look at
+`lyrics-fetcher-neteasecloud--format-query'.  If the search is
+successful, CALLBACK will be called with the result.
+
+If SYNC is non-nil, perform request synchronously and ask the
+user to pick the matching search result.
+
+When EDIT is non-nil, edit the query in minibuffer before search.
+Genius usually struggles to find song if there is extra
+information in the title.
+"
+  (lyrics-fetcher-neteasecloud--do-query
+   track
+   (lambda (data)
+     (lyrics-fetcher-neteasecloud--fetch-lyrics
+      (lyrics-fetcher-neteasecloud--get-song-id data sync)
+      callback
+      sync))
+   sync
+   edit))
+
+(defun lyrics-fetcher-neteasecloud--fetch-lyrics (song-id callback &optional sync)
+  "Fetch lyrics from 'music.163.com' page at URL and call CALLBACK with the result.
+
+If SYNC is non-nil, the request will be performed synchronously."
+  (message "Getting lyrics from NeteaseCloud API...")
+  (request
+    (format "http://music.163.com/api/song/lyric?id=%s&lv=1&kv=1&tv=-1" song-id)
+    :parser 'json-read
+    :sync sync
+    :success (cl-function
+     (lambda (&key data &allow-other-keys)
+       (funcall callback (alist-get 'lyric (alist-get 'lrc data)))))
+    :error
+    (cl-function
+     (lambda (&key error-thrown &allow-other-keys)
+       (message "Error!: %S" error-thrown)))
+    ))
```

Comment:
Please, don't leave trailing parentheses on distinct lines, it's an uncommon practice. E.g. see [bbatsov's elisp style guide](https://github.com/bbatsov/emacs-lisp-style-guide#source-code-layout--organization)